### PR TITLE
Adding example on learning graph structured data embedding in Hyperbolic manifold

### PR DESCRIPTION
**Pull request by tgeral68**
Works for numpy only (currently working to make it works for pytorch):
TODO:

- Adding an other loss
- Learn using retraction methods
- Use autograd instead of defining log_sigmoid grad
- Make it works an pytorch/tensorflow

Here a gif to show how embeddings are moving during learning:
learning_GSD

### CODE REVIEW

File: examples/learning_graph_structured_data_h2.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    """Logsigmoid function."""
+    return gs.log((1 / (1 + gs.exp(-x))))
+
+
+def grad_log_sigmoid(x):
+    """Gradient of logsigmoid function."""
+    return (1 / (1 + gs.exp(x)))
+
+
+def squared_distance(example_embedding, context_embedding,
```
Comment (ninamiolane) [inline examples/learning_graph_structured_data_h2.py:23]: RiemannianMetric has a `squared_dist` method implemented. Doesn't it work?

File: examples/learning_graph_structured_data_h2.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+                     manifold):
+    """Squared distance considering the manifold metric."""
+    return \
+        manifold.metric.dist(example_embedding, context_embedding)**2
+
+
+def grad_squared_distance(example_embedding, context_embedding,
```
Comment (ninamiolane) [inline examples/learning_graph_structured_data_h2.py:30]: This seems to be very close to what you want to do:
https://github.com/geomstats/geomstats/blob/05a2bcc2d696b56bc428c5d3d30e2be889920f4a/geomstats/geometry/riemannian_metric.py#L19

Is there anyway that you could *use* this function, or adding your function to hyperbolic.py if it very different?

Such functions are very useful even outside this example!

File: examples/learning_graph_structured_data_h2.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    return -2 * log_map
+
+
+def loss(example_embedding, context_embedding, negative_embedding, manifold):
+    """Compute loss and grad.
+
+    Compute loss and grad given embedding of the current example,
+    embedding of the context and negative sampling embedding.
+    """
+    N, D = negative_embedding.shape[0], example_embedding.shape[-1]
```
Comment (ninamiolane) [inline examples/learning_graph_structured_data_h2.py:45]: no uppercase letter for variables, no single-letter variables (sorry!!).

File: examples/learning_graph_structured_data_h2.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    max_epochs = 100
+    lr = .05
+    n_negative = 2
+    context_size = 1
+    karate_graph = gdp.Graph(
+        graph_matrix_path="examples/data/graph_karate/karate.txt",
+        labels_path="examples/data/graph_karate/karate_labels.txt")
+
+    nb_vertices_by_edges =\
+        [len(e_2) for e_1, e_2 in karate_graph.edges.items()]
```
Comment (ninamiolane) [inline examples/learning_graph_structured_data_h2.py:106]: `for _, e_2 in karate_graph.edges.items()`

File: examples/learning_graph_structured_data_h2.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    logging.info('Nb edges: %s' % len(karate_graph.edges))
```
Comment (ninamiolane) [inline examples/learning_graph_structured_data_h2.py:107]: No abbreviation in public prints, so that it is 100% clean?

File: examples/learning_graph_structured_data_h2.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+    negative_sampling_table = gs.array(negative_sampling_table)
+    random_walks = karate_graph.random_walk()
+    embeddings = gs.random.randn(karate_graph.n_nodes, dim)
+    embeddings = embeddings * 0.2
+
+    hyperbolic_manifold = Hyperbolic(2, coords_type='ball')
+    labels =\
+        [karate_graph.labels[i][0] for i in range(len(karate_graph.labels))]
+    colors = {1: "b", 2: "r"}
```
Comment (ninamiolane) [inline examples/learning_graph_structured_data_h2.py:126]: :D " -> '

File: examples/learning_graph_structured_data_h2.py [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    example_grad = -(positive_grad + negative_grad.sum(axis=0))
+
+    return total_loss, example_grad
+
+
+def main():
+    """Learning Poincaré graph embedding.
+
+    Learns Poincaré Ball embedding by using Riemannian
+    gradient descent algorithm
```
Comment (ninamiolane) [inline examples/learning_graph_structured_data_h2.py:94]: Missing .

Comment (ninamiolane) [reply examples/learning_graph_structured_data_h2.py:94]: Maybe you want to reference one of your papers here? (Using conventions for references in docstrings?)

File: examples/learning_graph_structured_data_h2.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+import matplotlib.pyplot as plt
+
+import geomstats.backend as gs
+import geomstats.visualization as visualization
+from geomstats.datasets import graph_data_preparation as gdp
+from geomstats.geometry.hyperbolic import Hyperbolic
+
+
+def log_sigmoid(x):
+    """Logsigmoid function."""
```
Comment (ninamiolane) [inline examples/learning_graph_structured_data_h2.py:14]: Missing parameters' descriptions in docstrings.

File: examples/learning_graph_structured_data_h2.py (hunk 2 above)
Comment (tgeral68) [reply examples/learning_graph_structured_data_h2.py:30]: Maybe renamed it in riemannian metric ? I do not know why this function is called loss instead of squaredDistance

Comment (tgeral68) [reply examples/learning_graph_structured_data_h2.py]: OK my next PR will be to refactor hyperbolic Space and Riemannian metrics.  I will make some changes to use  (and rename) upper-class methods.

Comment (ninamiolane) [reply examples/learning_graph_structured_data_h2.py]: This function is called `loss` because it is used in a deep learning example on SE(3) in the `geomstats/applications` folder.

File: geomstats/geometry/hyperbolic.py [hunk 9]
```diff
@@ -881,6 +882,29 @@ def dist(self, point_a, point_b):
             raise NotImplementedError(
                 'dist is only implemented for ball and extrinsic')
 
+    def squared_dist(self, point_a, point_b):
+        """Geodesic squared hyperbolic distance.
+
+        Redefinition the squared distance implemented
```
Comment (ninamiolane) [inline geomstats/geometry/hyperbolic.py:888]: Override the squared distance ... .

I don't understand why you need to override this method. The squared dist from RiemannianMetric is not called?

File: geomstats/geometry/hyperbolic.py [hunk 10]
```diff
@@ -881,6 +882,29 @@ def dist(self, point_a, point_b):
...
+            First point in hyperbolic space.
+        point_b : array-like, shape=[n_samples, dimension]
+            Second point in hyperbolic space.
+
+        Returns
+        -------
+        dist : array-like, shape=[n_samples, 1]
+            Geodesic squared distance between the two points.
+        """
+        if self.coords_type in ["intrinsic", "extrinsic"]:
```
Comment (ninamiolane) [inline geomstats/geometry/hyperbolic.py:903]: " -> '
